Notify websocket service on abnormal socket closure

bikram990 opened the pull request:
This pull request fixed issue #60

ianpartridge commented:
Thanks for your PR! Before we review it, could you please sign the CLA? (link above).  Also, your PR seems to have failed CI, can you take a look?

bikram990 commented:
Hi @ianpartridge,

I've fixed the CI build failure. While looking at it I found another issue. I've updated the issue #60 with same.

The CI build again failed due to a timeout issue.
This is the first time I'm submitting a PR to any opensource repository. Could you please help/guide me to fulfill the requirements for PR?

Thanks

bikram990 commented:
@Andrew-Lees11  Did you get a change to review this PR?

Andrew-Lees11 commented:
Hello @bikram990
Sorry for the delay. I have been looking over this PR today. Firstly thank you for raising the issue and PR. You have caught some good bugs. There are currently two things with your PR I'm looking into:
1. You have added a test which is good. The test passes even without your changes though so It isn't covering your case
2. connectionClosed calls processor?.close(). This would cause a loop if you didn't set the processor as nil and I wanted to check if there is a better way to do this.

Could you expand on what is happening when you get your issue? Are you using master branch with a connection timeout or latest release? Have you tested this PR in your case and did it fix your issue? 

Finally could you either rebase or merge your changes on master.
I apologize for all the question but I want to make sure I understand this change before I merge it.

bikram990 commented:
> You have added a test which is good. The test passes even without your changes though so It isn't covering your problem case.

I've not added any new test case, I've modified the existing test cases. One issue with test case was that the connection was closed abnormally and it was expecting to get a noReason error code. Second issue with test case was specific to macOS. In Xocde the tests were giving false results even though test was failed. Third issue is related to timeout. The timeout issue may be because of network issues on the build server or no delay before closing the socket. I've not fixed the timeout issue But would like to have a fix for the same.

> connectionClosed calls processor?.close(). This would cause a loop if you didn't set the processor as nil and I wanted to check if there is a better way to do this.

I just followed the same pattern used in Kitura-Net. Personally I would add a new delegate method for this. But for this PR I wanted to keep my modifications as minimum as possible.

> Could you expand on what is happening when you get your issue? Are you using master branch with a connection timeout or latest release? Have you tested this PR in your case and did it fix your issue?

For my project, I'm using master branch without any timeout. The client for my server was killed forcefully and my server was never notified for the same. As my server was not notified, the server was still trying to send messages to client.

Andrew-Lees11 commented:
Thank you for your comprehensive response.
We have seen the timeout issue on travis before and I'm not to worried about that making it fail. 

Looking at the code a agree that a delegate would probably be better but your solution will ensure that if the close function is called the service.disconnected should also be called.

Have you run your project with your branch? If the client never notifies the server then i'm not sure close will be called. Setting a connection timeout makes the server clean up connections that are not responding and should call service?.disconnected so your service is aware the client has gone.

bikram990 commented:
@Andrew-Lees11  Yes, I've run my project with above changes and it works. Basically in Kitura-Net when a socket is open it creates a dispatch read source. This dispatch source gets a cancel message when ever a socket is closed abnormally. In implementation of cancel it closes the socket from server side and notifies its delegate (WebSocketProcessor) about close.

I agree a timeout will make the server clean up stale connections. But for timeout period server is using more resources that it should. The above change makes it event based.

Andrew-Lees11 commented on Sources/KituraWebSocket/WSSocketProcessor.swift:
```diff
@@ -108,6 +108,8 @@ class WSSocketProcessor: IncomingSocketProcessor {
     /// Close the socket and mark this handler as no longer in progress.
     public func close() {
         handler?.prepareToClose()
+        connection.processor = nil
```
So I killing the processor here means it can be closed before prepareToClose completes. Instead I suggest:
1. make `active` in `connection` internal,
2. replace these lines with:
```swift
if connection.active {
    connection.connectionClosed(reason: .closedAbnormally)
}
```
above `handler?.prepareToClose()`

bikram990 replied:
Actually `active` is a `private` variable in `WebSocketConnection`. So I coded the other way as I didn't want to change access levels for any variable.

Above changes makes sense. I'll push these changes.

bikram990 replied:
@Andrew-Lees11  Another issue just came to my mind after pushing the suggested changes. The issue is that connection is holding a strong reference to processor and processor is holding a strong reference to connection. It becomes a cyclic reference and these objects will never be released.

I'm not expert in Swift. Could you please validate my above theory?

bikram990 replied:
I missed the part that processor is a `weak`. Thanks @Andrew-Lees11

Andrew-Lees11 commented on Tests/KituraWebSocketTests/TestWebSocketService.swift:
```diff
@@ -83,7 +83,8 @@ class TestWebSocketService: WebSocketService {
         XCTAssertNotNil(connections[connection.id], "Client ID from connect wasn't client ID from disconnect")
-        XCTAssertEqual(Int(closeReason.code()), Int(reason.code()), "Excpected close reason code of \(closeReason) received \(reason)")
+        XCTAssert((closeReason.code() == reason.code()), "Excpected close reason code of \(closeReason) received \(reason)")
+        //XCTAssertEqual was giving false success cases
```
This is a good catch but could you remove this comment before you merge.

bikram990 replied:
Sure

Andrew-Lees11 commented:
Could you also add the following to the connection cleanup tests:
```
("testProccessorClose", testProccessorClose),
```
```
    func testProccessorClose() {
        let service = register(closeReason: .closedAbnormally, connectionTimeout: nil)
        
        performServerTest() { expectation in
            XCTAssertEqual(service.connections.count, 0, "Connections left on service at start of test")
            guard let socket = self.sendUpgradeRequest(toPath: self.servicePath, usingKey: self.secWebKey) else { return }
            let _ = self.checkUpgradeResponse(from: socket, forKey: self.secWebKey)
            usleep(1500)
            XCTAssertEqual(service.connections.count, 1, "Failed to create connection to service")
            let connections = Array(service.connections.values)
            connections[0].processor?.close()
            usleep(1500)
            XCTAssertEqual(service.connections.count, 0, "Service was not notified of connection disconnect")
            socket.close()
            usleep(150)
            expectation.fulfill()
        }
    }
```

bikram990 commented:
@Andrew-Lees11 I've made the suggested changes. Could you please review?

pushkarnk commented:
I agree with the proposed fix.
